### DISCUSSION TITLE
Optional parameter default validation behavior change between 0.112.3 and 0.112.4

### DISCUSSION THREAD

**Discussion by oliver-rew** [fastapi/fastapi#12332] [Questions]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from fastapi import FastAPI, Query
import uvicorn

app = FastAPI()

@app.get("/")
def get(foo: str = Query(min_length=1, default="")):
    return {"foo": foo}

uvicorn.run(app, port=5001)
```

### Description

We noticed a recent change in optional query parameter validation behavior and I narrowed it down to a change between versions 0.112.3 and 0.112.4, and it is still present in the latest 0.115.0.

When an optional query param is defined with `default=""` and `min_length=0`, and it is not specified in the request, previously it would be assigned the default, the `min_length` validation would be skipped, and the endpoint function would run. Now, it seems the validation is run on the default value, which can fail.

Example (example code in main.py):
**<=0.112.3**
```
pip install fastapi==0.112.3
python3 main.py
...
INFO:     127.0.0.1:59103 - "GET / HTTP/1.1" 200 OK
```
```
curl "localhost:5001"
{"foo":""}
```

**>=0.112.4**
```
pip install fastapi==0.112.4
python3 main.py
INFO:     127.0.0.1:59163 - "GET / HTTP/1.1" 422 Unprocessable Entity
```
```
curl "localhost:5001"
{"detail":[{"type":"string_too_short","loc":["query","foo"],"msg":"String should have at least 1 character","input":"","ctx":{"min_length":1}}]}
```

I briefly diffed the 2 tags and I would guess the issue lies in recent changes to `request_params_to_args(` here:
https://github.com/fastapi/fastapi/blob/847296e885ed83bc333b6cc0a3000d6242083b87/fastapi/dependencies/utils.py#L730

In the past, it looks like unspecified params were assigned the default value and field validation was skipped, but now the validation is performed on the default value.

TBH the new behavior doesn't seem terribly unreasonable, I couldn't find anywhere in the docs that specify whether `Query()` validations are performed on the default values. Indeed, looking back on our code, it does feel weird to have a default value that doesn't pass the validations, though we probably left it because it worked at the time. However, I didn't see mention of the change in the changelog and it did break some functionality on our side, so I figured I would bring it up.

Thanks for all the hard work, FastAPI rocks! 🥇 

### Operating System

Linux

### Operating System Details

Observed on both linux and mac

### FastAPI Version

0.112.4

### Pydantic Version

2.9.2

### Python Version

Python 3.12.3

### Additional Context

no, please see description and example

**Comment by sinisaos** [comment]
@oliver-rew It's tricky. I'm sorry if I'm wrong, but I think the main problem is that FastAPI treats an empty string as a `None` value, but that's not true. The `None` keyword is used to define a `null value` or `no value at all`. `None` is not the same as an `empty string`. I think that in version `<=0.112.3`, in the `request_params_to_args` method, the condition was wrong and that's why it passed validation if the default value was an empty string. 
```python
# version 0.112.3
# this was in dependencies/utils.py
for field in required_params:
    if is_scalar_sequence_field(field) and isinstance(
        received_params, (QueryParams, Headers)
    ): # always False for default="" !!!!!
        value = received_params.getlist(field.alias) or field.default
    else:
        value = received_params.get(field.alias)
```
I think this piece of code should have been like this

```python
# version 0.112.3
# this was in dependencies/utils.py
for field in required_params:
    if is_scalar_sequence_field(field) and isinstance(
        received_params, (QueryParams, Headers)
    ):
        value = received_params.getlist(field.alias) or field.default
    else:
        value = received_params.get(field.alias) or field.default
```
and then, even in that version, the validation wouldn't pass, because later the value is compared to `None` (it was always `None` because the condition for `default=""` was always `False`), not the `empty string`.

Although it makes perfect sense to me that if the default value is an empty string (`len("") == 0`) and is less than `min_length`, the validation should not pass. I'm not sure what the best solution is, but if you still want an empty string as the default and want your code to pass validation, patch your FastAPI code with an empty string check along with `None` check. Replace [this line](https://github.com/fastapi/fastapi/blob/847296e885ed83bc333b6cc0a3000d6242083b87/fastapi/dependencies/utils.py#L691) with this
```python
if value is None or not value: # check empty string along None
```
Result is:
```bash
> curl http://localhost:8000
{"foo":""}
```
Sorry for really long comment, hope that make sense.

**Reply by oliver-rew** [reply]
@sinisaos thanks for the response. We will probably just remove the `min_length` validation. TBH it isn't 100% critical and when we noticed it after this issue, it seemed a bit off with `default=""`. I just wanted to bring the change to your attention. Thanks!

**Comment by oliver-rew** [comment]
I was looking through this code and noticed that `pydantic.fields.Field`, which is wrapped by `fastapi.ModelField`, has a `validate_default` default option. Would it be reasonable for FastAPI to abide by this in `request_params_to_args` as well?
https://docs.pydantic.dev/latest/api/fields/#pydantic.fields.Field

**Comment by sinisaos** [comment]
@oliver-rew That seems reasonable, but you have to experiment with it. I think the easiest way to achieve the same behavior as you have before is to check an `empty string` along with `None` as I wrote in the previous comment. Good luck.

**Comment by DeoLeung** [comment]
it seems fastapi would validate the default value after that version. it will be great if we have some global toggle for that, since we generally don't capture 422 error, if my test not covering the default value case , it may be a bit hard to discover it later before user/frontend complains :(
